File: src/main/java/com/project/ems/service/LeaveService.java
package com.project.ems.service;

import com.project.ems.dto.LeaveApplyRequest;
import com.project.ems.dto.LeaveResponse;
import com.project.ems.dto.LeaveReviewRequest;

import java.util.List;

/**
 * Service contract for leave-management operations.
 */
public interface LeaveService {

    /**
     * Submits a leave request for the authenticated user.
     *
     * @param username authenticated username
     * @param request leave details
     * @return submitted leave application
     */
    LeaveResponse applyLeave(
            String username,
            LeaveApplyRequest request
    );

    /**
     * Returns leave applications submitted by the authenticated user.
     *
     * @param username authenticated username
     * @return user's leave history
     */
    List<LeaveResponse> getMyLeaves(String username);

    /**
     * Returns all leave applications.
     *
     * Intended for ADMIN and HR users.
     *
     * @return all leave applications
     */
    List<LeaveResponse> getAllLeaves();

    /**
     * Approves a pending leave application.
     *
     * @param leaveId leave application ID
     * @param reviewerUsername ADMIN or HR username
     * @param request optional review comment
     * @return approved leave application
     */
    LeaveResponse approveLeave(
            Long leaveId,
            String reviewerUsername,
            LeaveReviewRequest request
    );

    /**
     * Rejects a pending leave application.
     *
     * @param leaveId leave application ID
     * @param reviewerUsername ADMIN or HR username
     * @param request optional review comment
     * @return rejected leave application
     */
    LeaveResponse rejectLeave(
            Long leaveId,
            String reviewerUsername,
            LeaveReviewRequest request
    );
}

File: frontend/src/leaveService.ts
import api from "./api";
import type {
  ApiResponse,
  LeaveApplyRequest,
  LeaveResponse,
  LeaveReviewRequest,
  LeaveStatus,
} from "./types";

/**
 * Submits a leave application for the authenticated user.
 *
 * Backend endpoint:
 * POST /api/v1/leaves
 */
export async function applyForLeave(
  request: LeaveApplyRequest,
): Promise<LeaveResponse> {
  const response = await api.post<
    ApiResponse<LeaveResponse>
  >("/leaves", request);

  return response.data.data;
}

/**
 * Retrieves leave applications belonging to the
 * currently authenticated user.
 *
 * Backend endpoint:
 * GET /api/v1/leaves/my
 */
export async function getMyLeaves(): Promise<
  LeaveResponse[]
> {
  const response = await api.get<
    ApiResponse<LeaveResponse[]>
  >("/leaves/my");

  return response.data.data;
}

/**
 * Retrieves every leave application.
 *
 * The backend permits only ADMIN and HR users.
 *
 * Backend endpoint:
 * GET /api/v1/leaves
 */
export async function getAllLeaves(): Promise<
  LeaveResponse[]
> {
  const response = await api.get<
    ApiResponse<LeaveResponse[]>
  >("/leaves");

  return response.data.data;
}

/**
 * Approves or rejects a pending leave application.
 */
export async function reviewLeave(
  leaveId: number,
  status: Extract<LeaveStatus, "APPROVED" | "REJECTED">,
  comment: string | null,
): Promise<LeaveResponse> {
  const action =
    status === "APPROVED" ? "approve" : "reject";

  const request: LeaveReviewRequest = {
    comment,
  };

  const response = await api.put<
    ApiResponse<LeaveResponse>
  >(`/leaves/${leaveId}/${action}`, request);

  return response.data.data;
}